coderd/keyrotate/rotate.go

```diff
@@ -0,0 +1,287 @@
+package keyrotate
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"encoding/hex"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/quartz"
+)
+
+const (
+	WorkspaceAppsTokenDuration = time.Minute
+	OIDCConvertTokenDuration   = time.Minute * 5
+	TailnetResumeTokenDuration = time.Hour * 24
+
+	// DefaultRotationInterval is the default interval at which keys are checked for rotation.
+	DefaultRotationInterval = time.Minute * 10
+	// DefaultKeyDuration is the default duration for which a key is valid. It applies to all features.
+	DefaultKeyDuration = time.Hour * 24 * 30
+)
+
+// Rotator is responsible for rotating keys in the database.
+type Rotator struct {
+	db          database.Store
+	logger      slog.Logger
+	clock       quartz.Clock
+	keyDuration time.Duration
+	// resultsCh is purely for testing.
+	resultsCh chan []database.CryptoKey
+
+	// The following fields are instantiated in Open.
+	ticker   *quartz.Ticker
+	features []database.CryptoKeyFeature
+}
+
+// Open instantiates a new Rotator. It ensures there's at least one
+// valid key per feature prior to returning. Close should be called
+// to ensure the ticker that is instantiated is not leaked.
+func Open(ctx context.Context, db database.Store, logger slog.Logger, clock quartz.Clock, rotateInterval time.Duration, keyDuration time.Duration, results chan []database.CryptoKey) (*Rotator, error) {
+	if keyDuration == 0 || rotateInterval == 0 {
+		return nil, xerrors.Errorf("key duration and rotate interval must be set")
+	}
+
+	kr := &Rotator{
+		db:          db,
+		keyDuration: keyDuration,
+		logger:      logger,
+		clock:       clock,
+		resultsCh:   results,
+
+		features: database.AllCryptoKeyFeatureValues(),
+		ticker:   clock.NewTicker(rotateInterval),
+	}
+	_, err := kr.rotateKeys(ctx)
+	if err != nil {
+		kr.Close()
+		return nil, xerrors.Errorf("rotate keys: %w", err)
+	}
+
+	return kr, nil
+}
+
+// Start begins the rotation routine. Callers should invoke this in a goroutine.
+func (k *Rotator) Start(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-k.ticker.C:
+		}
+
+		modifiedKeys, err := k.rotateKeys(ctx)
+		if err != nil {
+			k.logger.Error(ctx, "failed to rotate keys", slog.Error(err))
+		}
+
+		// This should only be called in test code so we don't
+		// bother to select on the push.
+		if k.resultsCh != nil {
+			k.resultsCh <- modifiedKeys
+		}
+	}
+}
+
+// rotateKeys checks for any keys needing rotation or deletion and
+// may insert a new key if it detects that a valid one does
+// not exist for a feature.
+func (k *Rotator) rotateKeys(ctx context.Context) ([]database.CryptoKey, error) {
+	var modifiedKeys []database.CryptoKey
+	return modifiedKeys, database.ReadModifyUpdate(k.db, func(tx database.Store) error {
```

Comment:
Might as well also grab a lock for this too to avoid rollbacks at all

Reply:
Why should we be concerned about not rolling back?

Reply:
What's the problem with grabbing a lock? We already have all of the code for it

Reply:
IDK it's probably fine it just feels a little heavy handed to me